### DISCUSSION TITLE
Combine support

### DISCUSSION THREAD

**Discussion by ancaarhip** [skiptools/skip#148] [Q&A] [answered]
Does skip support combine?

**Comment by aabewhite** [comment, marked as answer]
Skip supports a minimal subset of Combine in the SkipModel framework. You can see what API is supported here:
https://github.com/skiptools/skip-model?tab=readme-ov-file#api-support

Please let us know if this does not meet your needs!

**Reply by ancaarhip** [reply]
I'm trying to transpile a pretty small app with Skip and one of the many errors I get is: "Unresolved reference: AnyPublisher"
Am I missing anything?

**Reply by aabewhite** [reply]
AnyPublisher is not implemented. It wouldn't be hard to add. Or can your use case use "any Publisher"? What exactly is the code?

If you haven't already seen it, the documentation has a section on migrating existing apps: 

https://skip.tools/docs/gettingstarted/#existing_development

**Reply by rj-atkinson** [reply]
Just stumbled upon this project, amazing initiative, exact dev pipeline I've been waiting for, would happily be a paying customer. 

However, my app depends on Combine and lots of AnyPublishers to manage multiple streams of realtime data. In my experience, moving to `any Publisher` fails frequently, which seems to be confirmed here:

https://forums.swift.org/t/is-there-any-advantage-in-using-anypublisher-with-swift-5-7/60587/7

If AnyPublisher isn't hard to add, that would be a huge roadblock removed for me.

**Reply by aabewhite** [reply]
Can you provide a concrete example of your publisher use? That will help us analyze whether the general pattern would be difficult to support.

**Reply by rj-atkinson** [reply]
Yea the link has the example but I'll paste it here for convenience:

```
let p: any Publisher<Int, Never> = Just(42)
p.map { $0 + 1 }
```

`❌ Member 'map' cannot be used on value of type 'any Publisher<Int, Never>'; consider using a generic constraint instead`

**Reply by aabewhite** [reply]
Sorry I didn't mean the "any Publisher" thing. I mean a real example that shows how you tend to use Publishers in your app. Both the production and consumption sides. Often people are using a lot of API surface without really being aware of it

**Reply by rj-atkinson** [reply]
Ah I see, used to the simple example being best but makes sense here.

See below for a quick pseudo version of the production code to show the surface area.

I'd be happy to work around missing pieces like `compactMap` and `switchToLatest` per the support listed here: https://skip.tools/docs/modules/skip-model/. The deal breaker is that I can no longer use Protocols and publishers together.



```
// Handles bluetooth comms
class BTManager: NSObject {
    var didUpdateValueForCharacteristicSubject: PassthroughSubject<CBCharacteristic, Never> = .init()
    var peripheralConnectedSubject: PassthroughSubject<(CBPeripheral, CBPeripheralState), Never> = .init()
}

// All peripheral devices conform
protocol PeripheralProtocol: ObservableObject {
     // switching this to any Publisher<Int16?, Never> is where things fall apart.
    var valuePublisher: AnyPublisher<Int16?, Never> { get }
}


// Many instances of this, PeripheralManagerB, C, D
final class PeripheralManagerA: PeripheralProtocol {
    public lazy var valuePublisher: AnyPublisher<Int16?, Never> = btManager.didUpdateValueForCharacteristicSubject
        .compactMap { $0.value }
        .map(PeripheralAFirmwareUtils.bytesToValue)
        .eraseToAnyPublisher()
}

// Manages active stream of data
class ActivePeripheralManager: ObservableObject {
    public lazy var valuePublisherSubject: PassthroughSubject<AnyPublisher<Int16?, Never>, Never> = .init()
    public lazy var valuePublisher: AnyPublisher<Int16?, Never> = valuePublisherSubject
        .switchToLatest()
        .eraseToAnyPublisher()
    
    public lazy var activePeripheralManagerPublisher: AnyPublisher<(any PeripheralProtocol)?, Never> = btManager.peripheralConnectedSubject
        .map { btManager.from(peripheral: $0)?.manager }
        .eraseToAnyPublisher()
    
    init() {
        activePeripheralManagerPublisher
            .compactMap { $0 }
            .sink { [weak self] in
                self?.valuePublisherSubject.send($0.valuePublisher)
            }
    }
}

// Any view consumes the active value
class SomeViewModel: ObservableObject {
    @Published var value: Int16?
    
    init() {
        activePeripheralManager.valuePublisher
            .assign(to: \.value, on: self)
    }
}
```

**Reply by aabewhite** [reply]
Thanks! And thanks for your patience. I think we can accommodate this. I'll try to get to it by the end of the week!

**Reply by rj-atkinson** [reply]
Awesome, thank you! Excited to dive in.

**Reply by aabewhite** [reply]
skip-model 0.8.1 includes AnyPublisher.init(_: Publisher) and Publisher.eraseToAnyPublisher().

Please let us know if you run into any issues!

Also note that if you want functions like compactMap, they're not hard to implement. You can use the implementations of other functions like map() as examples. For Publishers we've chosen to do the implementation in Kotlin, so the summary is that you stub out the new functions in Publisher.swift and then implement them in Publisher.kt. Kotlin is enough like Swift that with the help of the existing examples, you'll likely be able to implement additional functionality even if you don't have experience with the language.

As a final step, add a unit test case and run the test suite against the Mac (only running the tests on the Mac will transpile, build, and run on Android too)

**Reply by rj-atkinson** [reply]
Really appreciate the responsiveness! It goes a long way when committing to a platform like this.

Good to know re: .compactMap, happy to contribute. I am looking to build a new onboarding flow for my iOS-only app in a few weeks and I'm hoping to start fresh with code that I can deploy to both platforms, while slowly migrating the rest for Android compatability.

**Reply by aabewhite** [reply]
Sounds great! I recommend joining our Slack when you get started. It's the easiest way to get help: https://skip.tools/slack/

**Comment by ancaarhip** [comment]
If I use "any Publisher" I'll need a different implementation for eraseToAnyPublisher
Then there is the Fail publisher the doesn't get resolved

Ex: 
func get<T: Codable>(_ url: URL) -> AnyPublisher<T, Error> {
        
        let cancellable = URLSession.shared.dataTaskPublisher(for: url)
            .map(\.data)
            .decode(type: T.self, decoder: JSONDecoder())
            .eraseToAnyPublisher()
        
        return Fail(error: NetworkError.invalidURL).eraseToAnyPublisher()
    }

**Reply by aabewhite** [reply]
I'm afraid this pattern is going to fail for several other reasons as well. You would find writing cross-platform Swift more frictionless by moving from Combine to async functions
